### DISCUSSION TITLE
Doge Testnet Tooling: Free TESTDOGE Faucet + Full History Electrs API + Block Explorer + Web Wallet

### DISCUSSION THREAD

**Discussion by cf** [dogecoin/dogecoin#3626] [Show and tell]
Interest in building on Doge has been exploding in recent months 🎉, so we decided to launch some community projects to get folks up and running on testnet with the goal being for anyone to be able to easily build on the doge testnet without spinning up their own node/infra.

## Doge Testnet Faucet
A free and easy to use faucet for TESTNET Doge, stocked with an initial donation of [6,969,420.69696969 TESTDOGE from QED](https://doge-testnet-explorer.qed.me/tx/b61a3de2b9805f922cb764342691758340087c51269f2779b3df55ec62978e85)
* Link: https://faucet.doge.toys
<a href="https://faucet.doge.toys/" target="_blank">
<img src="https://github.com/user-attachments/assets/a355756d-c634-450c-b620-980149b45e73" width="400" />
</a>

## Public Electrs API and Block Explorer
A public full history API for the testnet (compatible with [doge-sdk](https://github.com/QEDProtocol/doge-sdk) + a blockstream flavored block explorer. To make it easy for integrations, the API supports `Access-Control-Allow-Origin: *` / CORS.
* API: https://doge-electrs-testnet-demo.qed.me
* Block Explorer: https://doge-testnet-explorer.qed.me
<a href="https://doge-testnet-explorer.qed.me/" target="_blank">
<img src="https://github.com/user-attachments/assets/355370cf-2267-401f-a2bb-cb781ee88099" width="400" />
</a>

#### Some Example API Calls
- Get Latest Blocks
	- GET `https://doge-electrs-testnet-demo.qed.me/blocks`
- Get Balance/UTXOs/Transactions for an Address
	- (Balance) `https://doge-electrs-testnet-demo.qed.me/address/nawiiYr7RC9UTbnJQz9riZn9pHocNJvAWB`
	- (UTXOs) GET `https://doge-electrs-testnet-demo.qed.me/address/nawiiYr7RC9UTbnJQz9riZn9pHocNJvAWB/utxo`
	- (Transactions) GET `https://doge-electrs-testnet-demo.qed.me/address/nawiiYr7RC9UTbnJQz9riZn9pHocNJvAWB/txs`
- Broadcast/Submit Transaction
	- POST `https://doge-electrs-testnet-demo.qed.me/tx <hex encoded transaction>`
	- Example: `curl 'https://doge-electrs-testnet-demo.qed.me/tx' --data-raw '<hex encoded transaction>'`

*An exhaustive list of APIs can be found [here](https://github.com/blockstream/esplora/blob/master/API.md)*

## Testnet Web Wallet
A fully featured web wallet for sending/receiving TESTDOGE with support for:
* Easy wallet creation + import/export via WIF and BIP39 mnemonic secrets
* Live balances and transactions feed
* Multi-wallet switching for fast testing

Link: https://testnet-wallet.doge.toys/
<a href="https://testnet-wallet.doge.toys/" target="_blank">
<img src="https://github.com/user-attachments/assets/08fb456b-d128-48a3-836e-29415131c6b2" height="400" />
</a>

In addition, we have dedicated a relatively beefy server to mining doge testnet blocks, so if your building on testnet you can be sure that blocks can be mined in a reasonable amount of time (1-2 minutes).

If anyone has any other tools they wish to see built for testnet, please let us know!

<3

**Comment by gexiao** [comment]
Helps a lot. Thank you!

**Comment by Dreiev** [comment]
Faucet not working
